pkg/core/template/engine.go

```diff
@@ -1,98 +1,162 @@
 package template
 
 import (
+	"encoding/json"
+	"fmt"
 	"sort"
 
 	"golang.org/x/text/language"
 
+	"github.com/iawaknahc/gomessageformat"
 	"github.com/skygeario/skygear-server/pkg/core/config"
 )
 
 type ResolveOptions struct {
-	Required bool
-	Key      string
+	Key string
+}
+
+type resolveResult struct {
+	Spec         Spec
+	TemplateBody string
+	// Translations is key -> tag -> translation.
+	// For example,
+	// {
+	//   "key1": {
+	//     "": "Hello",
+	//     "en": "Hello",
+	//     "en-US": "Hi!",
+	//     "zh": "你好"
+	//   }
+	// }
+	Translations map[string]map[string]string
```

Comment:
Would it be better to flatten it like this?
```json
{
    "key1": "Hello",
    "key1#en": "Hello",
    "key1#en-US": "Hi!",
    "key1#zh":  "你好"
}
```

In this ways, we can also resolve the other form of localization in configuration.

Reply:
First it would impose a new restriction on the key naming. Key cannot have `#` sign otherwise it would be problematic.

Second the two-map approach does not require extra key manipulation to index. It is simpler.

Third the OIDC style localization is not yet done right now. I am going to support it in another PR (because your configuration refactoring introduces some configs that are not client metadata but use OIDC style localization, so I didn't do OIDC style localization in this PR). I am still thinking whether we should reuse MessageFormat for OIDC style localization. MessageFormat is powerful when we give it arguments to deal with plural and gender.

Reply:
> I am still thinking whether we should reuse MessageFormat for OIDC style localization. MessageFormat is powerful when we give it arguments to deal with plural and gender.

I don't think we need to use it, since localization in configuration should be static strings without arguments, and I don't the OAuth/OIDC spec allows it.

Reply:
> I don't think we need to use it, since localization in configuration should be static strings without arguments, and I don't the OAuth/OIDC spec allows it.

And that's why I don't think flattening the map here is helpful. The two localization processes are different. The data structure they use can also be different.